scripts/fit_pgd.py

```diff
@@ -0,0 +1,165 @@
+from functools import partial
+import flowpm.scipy.interpolate as interpolate
+
+from nbodykit.cosmology import Cosmology
+from nbodykit.cosmology.power.halofit import HalofitPower
+from astropy.cosmology import Planck15
+import astropy.units as u
+
+flags.DEFINE_string("filename", "results_fit_PGD.pkl", "Output filename")
+flags.DEFINE_integer("niter", 51, "Number of iterations of the first PGD fit")
+flags.DEFINE_integer("niter_refine", 11,
+                     "Number of iterations of subsequent PGD fit")
+flags.DEFINE_float("learning_rate", 0.1, "ADAM learning rate for the PGD optim")
+flags.DEFINE_integer("batch_size", 8, "Number of random draws")
+
+flags.DEFINE_float("alpha0", 0.3, "Initial guess for alpha at z=0")
+flags.DEFINE_float("kl0", 0.3, "Initial guess for kl at z=0")
+flags.DEFINE_float("ks0", 10, "Initial guess for ks at z=0")
+
+flags.DEFINE_boolean("fix_scales", False,
+                     "Whether to fix the scales after the initial fit at z=0")
+
+flags.DEFINE_float("a_init", 0.1, "Initial scale factor")
+flags.DEFINE_integer("nsteps", 40, "Number of steps in the N-body simulation")
+flags.DEFINE_float("box_size", 64.,
+                   "Transverse comoving size of the simulation volume")
+flags.DEFINE_integer("nc", 64,
+                     "Number of transverse voxels in the simulation volume")
+flags.DEFINE_integer("B", 1, "Scale resolution factor")
+
+FLAGS = flags.FLAGS
+
+
+def pgd_loss(alpha, scales, state, target_pk, return_pk=False):
+  """
+  Defines the loss function for the PGD parameters
+  """
+  shape = state.get_shape()
+  batch_size = shape[1]
+  pgdparams = tf.concat([alpha, scales], 0)
+
+  # Step I: Apply PGD to the state vector
+  pdgized_state = state[0] + tfpm.PGD_correction(
+      state, pgdparams, nc=[FLAGS.nc] * 3, pm_nc_factor=FLAGS.B)
+
+  # Step II: Painting
+  field = cic_paint(tf.zeros([batch_size] + [FLAGS.nc] * 3), pdgized_state)
+
+  # Step III: Compute power spectrum
+  k, pk = flowpm.power_spectrum(
+      field,
+      boxsize=np.array([FLAGS.box_size] * 3),
+      kmin=np.pi / FLAGS.box_size,
+      dk=2 * np.pi / FLAGS.box_size)
+  # Averaging pk over realisations
+  pk = tf.reduce_mean(pk, axis=0)
+
+  # Step IV: compute loss
+  weight = 1 - k / (np.pi * FLAGS.nc / FLAGS.box_size)
+  weight = tf.convert_to_tensor(weight / weight.sum(), dtype=tf.float32)
+  rescale_factor = 1.0  # pk[0]/target_pk[0] # To account for variance on large scale
+  loss = tf.reduce_sum((weight * (1 - pk / target_pk / rescale_factor))**2)
+  if return_pk:
+    return loss, pk
+  else:
+    return loss
+
+
+def main(_):
+  cosmology = flowpm.cosmology.Planck15()
+  # Create a simple Planck15 cosmology without neutrinos, and makes sure sigma8
+  # is matched
+  nbdykit_cosmo = Cosmology.from_astropy(Planck15.clone(m_nu=0 * u.eV))
+  nbdykit_cosmo = nbdykit_cosmo.match(sigma8=cosmology.sigma8.numpy())
+
+  # Compute the k vectora that will be needed in the PGD fit
+  k, _ = flowpm.power_spectrum(
```

Comment:
Minor issue that can totally be ignored and this has nothing to do with PGD, but maybe we want to use _initialize_pk or rename that function to get the parameters which are not dependent on the field. 
For instance, in addition to just extracting k-vector, there have been occassions when I needed to extract bin-edges or the number of modes in a bin.

Reply:
sure.... but if not needed for this script, I would opt for simplicity. the _initialize_pk is meant to remain secret.